react/features/calendar-sync/middleware.js

```diff
@@ -0,0 +1,207 @@
+// @flow
+import Logger from 'jitsi-meet-logger';
+import RNCalendarEvents from 'react-native-calendar-events';
+
+import { SET_ROOM } from '../base/conference';
+import { MiddlewareRegistry } from '../base/redux';
+import { APPLINK_SCHEME, parseURIString } from '../base/util';
+
+import { APP_WILL_MOUNT } from '../app';
+
+import { maybeAddNewKnownDomain, updateCalendarEntryList } from './actions';
+import { REFRESH_CALENDAR_ENTRY_LIST } from './actionTypes';
+
+const FETCH_END_DAYS = 10;
+const FETCH_START_DAYS = -1;
+const MAX_LIST_LENGTH = 10;
+const logger = Logger.getLogger(__filename);
+
+MiddlewareRegistry.register(store => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+    case APP_WILL_MOUNT:
+        _ensureDefaultServer(store);
+        _fetchCalendarEntries(store);
+        break;
+    case REFRESH_CALENDAR_ENTRY_LIST:
+        _fetchCalendarEntries(store);
+        break;
+    case SET_ROOM:
+        _parseAndAddDomain(store);
+    }
+
+    return result;
+});
+
+/**
+ * Ensures calendar access if possible and resolves the promise if it's granted.
+ *
+ * @private
+ * @returns {Promise}
+ */
+function _ensureCalendarAccess() {
+    return new Promise((resolve, reject) => {
+        RNCalendarEvents.authorizationStatus()
+            .then(status => {
+                if (status === 'authorized') {
+                    resolve();
+                } else if (status === 'undetermined') {
+                    RNCalendarEvents.authorizeEventStore()
```

Comment:
Because there's some duplication with the 'authorized' checks, I'm wondering if it would be possible to refactor the promise flow so that 1) the wrapping in a Promise is not needed and 2) there is nesting of promises. I find a flat promise chain easier to read. One idea I can think of is checking if status is undetermined and returning authorizeEventStore() or the status again, and afterwards a chained .then doing a check for authorized. It sounds better to me in my mind but I haven't implemented it so it might look worse.

Reply:
Honestly I tried to rethink this logic and see how an easier solution could come out of it, but I didn't really manage to get to one. Can you please explain your idea a bit more?

Reply:
Nevermind, I think I now know what you mean. The permission handling has been changed a bit, you'll see, and I think it now does what you also meant

Reply:
Yeah, I tend to be a mysterious, inscrutable man. Thanks for tackling a refactor of the chain.